### ISSUE TITLE
ALPN not available

### ISSUE THREAD

**Issue by johndoesenior** [mitmproxy/mitmproxy#3095]
##### Steps to reproduce the problem:

1. running mitmproxy transparent mode with ssl on
2. going to any website (HTTPS)
results in failure

ubuntu 16.04
python 3.6

```
Error: 192.168.42.19:49597: Traceback (most recent call last):(more in eventlog)Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/proxy/server.py", line 121, in handle
    root_layer()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/proxy/modes/transparent_proxy.py", line 19, in __call__
    layer()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/proxy/protocol/tls.py", line 279, in __call__
    self._establish_tls_with_client_and_server()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/proxy/protocol/tls.py", line 355, in _establish_tls_with_client_and_server
    self._establish_tls_with_server()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/proxy/protocol/tls.py", line 444, in _establish_tls_with_server
    **args
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/connections.py", line 292, in establish_tls
    self.convert_to_tls(cert=client_cert, sni=sni, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/net/tcp.py", line 388, in convert_to_tls
    **sslctx_kwargs
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/net/tls.py", line 285, in create_client_context
    **sslctx_kwargs,
  File "/usr/local/lib/python3.6/site-packages/mitmproxy/net/tls.py", line 194, in _create_ssl_context
    context.set_alpn_protos(alpn_protos)
  File "/usr/local/lib/python3.6/site-packages/OpenSSL/SSL.py", line 654, in explode
    raise NotImplementedError(error)
NotImplementedError: ALPN not available

mitmproxy has crashed!
Please lodge a bug report at: https://github.com/mitmproxy/mitmproxy
```

Thanks

**Comment by Kriechi** [comment]
Which mitmproxy version are you using? How did you install it?

**Comment by johndoesenior** [comment]
Compiled from source. 3.0.4

**Comment by Kriechi** [comment]
Do you have a recent version of pyopenssl and cryptography installed?

**Comment by johndoesenior** [comment]
Will check early next week and update. thanks

**Comment by ulises2k** [comment]
Hello,
I have the same problem in raspberry pi 3. 
- Raspbian GNU/Linux 8 (Debian jessie)
- Python 3.6.4 (install from source code)

root@raspberrypi:~ # mitmproxy --version
Mitmproxy: 3.0.4
Python:    3.6.4
OpenSSL:   OpenSSL 1.0.1t  3 May 2016
Platform:  Linux-4.9.35-v7+-armv7l-with-debian-8.0

**Comment by Kriechi** [comment]
ALPN is available in openssl 1.0.2: https://www.openssl.org/news/openssl-1.0.2-notes.html
You need to upgrade your openssl and make sure all the packages are compiled/built against it.
